### DISCUSSION TITLE
Dynamic input fields

### DISCUSSION THREAD

**Discussion by pkese** [fsprojects/FSharp.Data.GraphQL#502] [Q&A]
I'm rendering my schema definitions (both inputs and outputs) dynamically, based on database schema.

For example if my database table has fields
```sql
CREATE TABLE my_table (
    id GUID NOT NULL
    name STRING
    age INT
    items ARRAY<INT>
)
```
... then I'm dynamically creating input filters that provide different functionality for each field depending on field type, e.g.
- `id` will have comparison but no `isNull` filters
- `name` will include string comparison options
- `age` will have numerical comparisons including `isNull`
- `items` is an array and may include `isEmpty`, `contains` filters.

```fsharp
let MyTableFilters = Define.InputObject("MyTableFilters", [
    Define.Input("id", NonNullGuidFilters)
    Define.Input("name", StringFilters)
    Define.Input("age", IntFilters)
    Define.Input("items", IntArrayFilters)
])
```

I can create such an input filters object, but I don't know how to extract values supplied in these input filters during the request processing.  
What kind of type should I define for such a case?

Documentation is not very helpful either.  
The page has an example of [defining an input object](https://fsprojects.github.io/FSharp.Data.GraphQL/type-system.html#Defining-an-Input-Object) without a concrete generic type but samples in actual code in repository all have generics:
```fsharp
let CreateAccountData = Define.InputObject("CreateAccountData", [ ... ])
// vs
let CreateAccountData = Define.InputObject<AccountData>("CreateAccountData", [ ... ])
```

So my question is:
1) where and how do I extract data from these input objects in the runtime?
2) what kind of a type should I supply it ... can it be :arrow_down: ?
```fsharp
let MyTableFilters = Define.InputObject<Dictionary<string,obj>>("MyTableFilters", [ ... ])
```

I'm currently maintaining a hacked version of the FSharp.Data.GraphQL codebase where I added support for
```fsharp
let MyTableFilters = Define.InputObject<System.Data.ExpandoObject>("MyTableFilters", [ ... ])
```
but now my project is getting some traction and I need to move away from my hackish solution that breaks on every merge from upstream.

**Comment by xperiandri** [comment]
It is better to implement a test for this. And if it fails I will pick up your code and make it work or propose changes if you want to implement yourself
